### DISCUSSION TITLE
Help with satellite setup with mympd

### DISCUSSION THREAD

**Discussion by GioF71** [jcorporation/myMPD#932] [Q&A]
Hello, 
I am using a blank issue because this is for sure not an *issue*. At most it might be a feature request. Anyway sorry for not using a template. 

First of all, kudos for the great work. I am using mympd and it is awesome.
However, I would like to ask your help in order to understand the best way to run mympd on a satellite mpd configuration.
I have mpd installed on multiple low powered SBCs like the Raspberry PI 3/4 and Asus Tinkerboard; mounting the library via smb/nfs from the nas and having to rebuild the library on each player is a pain. Also, it can be painfully slow with large libraries and in any case play statistics would be disseminated through all the mpd instances.

I currently get away from this scenario with a minimserver library and upmpdcli, as well as by using other solutions like Logitech Media Server.

But since I saw first MyMPD, I would really like to be able to use it also on those players.

Do you think this might be possible, or do you have a suggestion on how I could configure mympd for this purpose?

Thanks a lot, and again, congratulations for the great work!

**Comment by GioF71** [comment]
I just wanted to specify that I normally know how to setup satellite mode in mpd, but mympd wants mpd to be configured with database_mode set to 'simple' AFAIK. Please correct me if I am wrong!

Thank you

**Comment by jcorporation** [comment]
> First of all, kudos for the great work. I am using mympd and it is awesome.

Many thanks.

> I am using a blank issue because this is for sure not an issue. At most it might be a feature request. 

That's why discussion exists ;)

> , but mympd wants mpd to be configured with database_mode set to 'simple' AFAIK.

No, myMPD should also work with other database modes, but I never tested an other setup.

For your setup the mpd satellite mode could be the best, but MPD has some limitations in this mode:

- https://github.com/MusicPlayerDaemon/MPD/issues/848
- https://github.com/MusicPlayerDaemon/MPD/issues/1105

If you want to test a MPD satellite setup, I could support you for the myMPD side and I am willing to implement features to improve the user experience for this type of setup.

For the MPD side: you should open issues or comment existing issues.

**Comment by GioF71** [comment]
Hello, thank you for your answer. I commented on the mpd issues, hopefully the author will find some time to solve those issues.
When I tried mympd in satellite mode, the issue was that the album cache, apparently, could never finish loading. Maybe I have not been patient enough, and unfortunately I did not collect the logs. But I will try again!
As the music directory, I used webdav, via https, of course on the local network. I don't know if this is slower than nfs/samba (but I believe it is), but does this matter? MyMPD did not seem to be actually scanning the files, as far as I understood. So I reverted to a regular mpd configuration on my desktop (which has a local copy of the whole library) and dropped, for now, the setup of mympd on the 'satellites'.

Is there something specific that I can do in order to help to understand what (if any) changes are needed on mympd side?

Thanks a lot again!

**Comment by jcorporation** [comment]
Right, myMPD creates the albumcache by shifting through all songs from mpd. How big is your library? 

https://github.com/jcorporation/myMPD/issues/913

**Comment by GioF71** [comment]
The library is 182k items.
For me, MPD is not generally slow at scanning al those files on a half-decent pc with local files. I did not measure, but on my xeon 2660 v3 desktop the initial scan took a very acceptable amount of time, and the updates are blazing fast now.
On the SBCs OTOH with files on a remote share (I was using webdav for convenience) I aborted the scan.

**Reply by jcorporation** [reply]
Eventually it is better to run MPD with multiple partitions at the central share and only stream music to the sattelites.

**Comment by GioF71** [comment]
Uhm, yes I noticed the partition feature, but I never used it and I do not really understand how it might solve my issue... 
I have looked [here](https://mpd.readthedocs.io/en/stable/protocol.html#client-to-client) but I do not understand how it would work across multiple instances. From what I see the subscribe command accepts a name only, am I missing something?

Thank you

**Reply by jcorporation** [reply]
Run only one instance with multiple partitions and stream to your clients, e.g. with snapcast.

The topic of your link has nothing todo with that.

**Reply by GioF71** [reply]
Ah ok, that is something that I do sometimes via icecast or httpd outputs. Never tried snapcast yet.
It's a little different from a satellite setup.
I was wondering if maybe mpd logging (which writes to microsd) was slowing the scan process, it might be worth trying again with logging disabled, do you think this could make a difference?

p.s. I learned something about partitions from the Cantata interface. I see I can move outputs to the partition, like you are probably suggesting. Not a bad idea anyway, thank you

**Reply by jcorporation** [reply]
Logging to sd-card is a bad idea. I prefer logging to ram as dietpi or myMPDos does.

Icecast or http output works also with my proposal.

**Reply by GioF71** [reply]
Ok perfect. I will try with logging disabled. Streaming is something I already do, but yes, it's not the same thing :-)

Have you considered to integrate mympd with a source like for example Navidrome? It uses the subsonic api. It might be a wonderful solution :-)

**Reply by jcorporation** [reply]
No, myMPD is only designed for MPD.

**Reply by bonelifer** [reply]
@GioF71 If you want to make your mpd library available to subsonic clients there's [mpdsonic](https://github.com/pborzenkov/mpdsonic)

**Reply by GioF71** [reply]
Hello @bonelifer, thank you. I saw that in the past, but actually I was suggesting the opposite: navigating and play from a subsonic library using mympd.

But the author said in the post before that myMPD is only designed for MPD. I respect his decision, it's his software, plus it's a perfectly legit decision.

However since that request I have built a subsonic plugin for upmpdcli (see discussion [here](https://github.com/navidrome/navidrome/discussions/2324) and [here](https://github.com/GioF71/upmpdcli-docker/discussions/291), so I solved my issue.

**Comment by GioF71** [comment]
Hello, today I am trying (from my old laptop) with mpd and a proxy configuration. MyMpd seems to be creating the albumcache as far as I understand.
I am using webdav for the music directory in mpd, so I wonder how do I need to configure "Music Directory" in MyMPD?
Maybe it would be better to use some share with nfs or smb?

p.s. Just wanted to mention that my Pi4 could successfully build the library using webdav as the music directory. I disabled log and let it work overnight, so I am sure that it took for sure less than 10 hours. Unfortunately I cannot be more precise.

**Reply by jcorporation** [reply]
myMPD does not need a configured music directory, you can configure it to "none". Without a music directory myMPD does not support lyrics and the flexible albumart configuration, therefor it could be preferable to use a nfs or smb mount.

**Reply by GioF71** [reply]
Ok perfect, that is what I did. I will try to switch to nfs or smb as soon as possible.
I have to tell that the albumcache eventually became available. The only problem is that, after a restart, the process restarts and it takes a long time. But IIRC there is an already solved ticket here, about preserving the album cache across restarts? I will have a look at the latest issues!

**Reply by GioF71** [reply]
I think maybe this [one](https://github.com/jcorporation/myMPD/issues/913) solved this issue?
Thank you!

**Reply by jcorporation** [reply]
Right, myMPD v10.2.0 saves the album and sticker cache across restarts.

**Reply by GioF71** [reply]
Awesome. I will be eagerly waiting for that release!

**Comment by tsunulukai** [comment]
Here's how you could use the partition feature to solve your issue:

Let's say you have a central NAS with your music collection and mympd.
Next to that, you have a few low-powered embedded device connected to stereo systems throughout your house.

What you can do is use mpd's partition feature to output each audio "zone" by multicast/unicast pulseaudio streaming from the central device to each low-powered device.

You then get on the central myMPD interface the option to select to which output your player will be outputting:
![image](https://user-images.githubusercontent.com/4215150/216442499-35bfd400-5bde-43e3-8243-4b6289629791.png)

In this case, I can either stream to all rooms using the "Multiroom" output, which then uses pulseaudio multicast to stream a single synchronized audio flow to all receivers, or output to a single room, which then uses pulseaudio unicast to stream to the selected receiver.

If I want to play different audio/playlists to differents zones, I create a new partition
![image](https://user-images.githubusercontent.com/4215150/216443199-e3d5a463-d453-4af2-9abc-b846485be69e.png)

Move the outputs I want to link to that partition:
![image](https://user-images.githubusercontent.com/4215150/216443388-e78323e2-829e-47e7-9e66-8d7bbb154956.png)

Let's say "Attic"
![image](https://user-images.githubusercontent.com/4215150/216443582-487565f0-bba4-43e1-b3ba-a11c18d72f8b.png)

![image](https://user-images.githubusercontent.com/4215150/216443642-c7518cd3-f6b7-4e63-9931-4b7fb831c472.png)

Now, the partition Jane has its own player, its own playlist and its own output(s).


How do you configure that on the different systems ?


# On All systems:
## Have Pulse audio start as system service with the following pulseaudio systemd unit file:
```
# cat /lib/systemd/system/pulseaudio.service
[Unit]
Description=PulseAudio Daemon
After=network.target

[Install]
WantedBy=multi-user.target

[Service]
Type=simple
PrivateTmp=true
ExecStart=/usr/bin/pulseaudio --system --realtime --disallow-exit --no-cpu-limit --disable-shm
```



# On the Central System:

## Configure Pulseaudio Group Permissions correctly
```
# grep pulse /etc/group
audio:x:29:pulse
pulse:x:110:
pulse-access:x:111:root,mpd
```

## Configure PulseAudio to Stream audio over the Network by RTP (RealTime Transport Protocol)
```
# cat /etc/pulse/system.pa
# < ORIGINAL FILE CONTENT STRIPPED FOR BREVITY >
# Only extra lines differing from standard configuration are shown
.include /etc/pulse/rtp_server.pa
```

```
# cat /etc/pulse/rtp_server.pa
load-module module-null-sink sink_name=rtp format=s32be channels=2 rate=48000 sink_properties="device.description='MPD RTP Multicast Sink'"
load-module module-rtp-send source=rtp.monitor destination_ip=239.0.0.100
set-default-sink rtp

load-module module-null-sink sink_name=rtpm format=s32be channels=2 rate=48000 sink_properties="device.description='RTP MultimediaRoom'"
load-module module-rtp-send source=rtpm.monitor destination_ip=192.168.0.11

load-module module-null-sink sink_name=rtps format=s32be channels=2 rate=48000 sink_properties="device.description='RTP Saloon'"
load-module module-rtp-send source=rtps.monitor destination_ip=192.168.0.12

load-module module-null-sink sink_name=rtpa format=s32be channels=2 rate=48000 sink_properties="device.description='RTP Attic '"
load-module module-rtp-send source=rtpa.monitor destination_ip=192.168.0.13

load-module module-null-sink sink_name=rtpb format=s32be channels=2 rate=48000 sink_properties="device.description='RTP BathRoom'"
load-module module-rtp-send source=rtpb.monitor destination_ip=192.168.0.14
```

## Configure MPD to send audio output to the different possible RTP sinks
```
# cat /etc/mpd.conf
# Only the Audio Output part from mpd.conf is shown here
# Audio Output ################################################################

# Pulseaudio Output
audio_output {
    type        "pulse"
    name        "MultiRoom"
    sink        "rtp"
}

audio_output {
    type        "pulse"
    name        "Multimedia"
    sink        "rtpm"
}

audio_output {
    type        "pulse"
    name        "Saloon"
    sink        "rtps"
}

audio_output {
    type        "pulse"
    name        "Attic"
    sink        "rtpa"
}

audio_output {
    type        "pulse"
    name        "Bathroom"
    sink        "rtpb"
}

```


# On the satellite systems:
No need to install mpd/mympd; only pulseaudio is required,  as system service preferrably.

## Configure Pulseaudio Group Permissions correctly
```
# grep pulse /etc/group
audio:x:29:pulse
pulse:x:116:
pulse-access:x:117:root
```

## Configure PulseAudio to receive audio over the Network by RTP (RealTime Transport Protocol)
```
# cat /etc/pulse/system.pa
# < ORIGINAL FILE CONTENT STRIPPED FOR BREVITY >
# Only extra lines differing from standard configuration are shown
.include /etc/pulse/rtp.pa
```

### Specific config file in Multimedia Room Embedded device
```
# cat /etc/pulse/rtp.pa
### Load the RTP receiver module
load-module module-rtp-recv sap_address=239.0.0.100
load-module module-rtp-recv sap_address=192.168.0.11
```

### Specific config file in Saloon Embedded device
```
# cat /etc/pulse/rtp.pa
### Load the RTP receiver module
load-module module-rtp-recv sap_address=239.0.0.100
load-module module-rtp-recv sap_address=192.168.0.12
```

### Specific config file in Attic Embedded device
```
# cat /etc/pulse/rtp.pa
### Load the RTP receiver module
load-module module-rtp-recv sap_address=239.0.0.100
load-module module-rtp-recv sap_address=192.168.0.13
```

### Specific config file in Bathroom Embedded device
```
# cat /etc/pulse/rtp.pa
### Load the RTP receiver module
load-module module-rtp-recv sap_address=239.0.0.100
load-module module-rtp-recv sap_address=192.168.0.14
```


## CPU consumption on satellite devices
is very low due to the heavy lifting being done on the central node:
![image](https://user-images.githubusercontent.com/4215150/216446762-98787071-d244-47f9-882a-985c83fcaba2.png)
Example here with a Seagate Dockstar (a low powered ARM plug computer) being used as audio receiver.

All operations regarding the scanning of the music collection, the playlists, etc only happens on the central node.

**Reply by jcorporation** [reply]
Great! I think this is worth to move it to the documentation in the additional topics section.

**Reply by tsunulukai** [reply]
I'll work it up and make a pull request

**Reply by jcorporation** [reply]
Top, base it on master as it is a documentation enhancement.

**Reply by tsunulukai** [reply]
Here it is: https://github.com/jcorporation/myMPD/pull/965
I hope you're OK with the assets location.

**Reply by dgcampea** [reply]
@tsunulukai I've got a similar setup here, how does your MultiRoom (multicast to every receiver) setup fare in terms of latency and synchronization between different receivers? I get some occasional hiccups and occasionally the audio gets buffered and induces timing differences across the receivers.

**Reply by tsunulukai** [reply]
@dgcampea As far as I'm concerned, multicast streaming with pulseaudio works like a breeze.

To answer your two concerns:
- Latency:
There's a tiny bit of latency that could be problematic if you're looking to have your audio in sync with a video, a game or videoconference, but when in use with mpd for audio-only streaming it's perfect.
N.B.: you can configure the latency in the pulseaudio configuration file on the receivers, where you need to strike a balance between enough latency to avoid audio glitches should the streaming buffer run out of content to play due to network congestion, and keeping it low enough not to experience an annoying delay between operations in your audio player and it's effect in the speakers. On my setup, I've left the latency settings untouched (no idea what the default value is) with no issue during playback. (I'm running a lightweight debootstrapped debian 11.6 on raspberry pis/dockstar for the receivers).
If you want to experience with it, just play with the `latency_msec` parameter:
`load-module module-rtp-recv latency_msec=10 sap_address=239.0.0.100`

- Synchronization:
You can sometime experience a slightly out-of-sync audio output from the satellite receivers when they're not properly synced. It usually happens when you just powered on the satellite receivers and begin the audio streaming immediately, before they can properly sync their clocks by ntp. In my setup, I only ever experience very light desychonisation that never persists beyond a minute after power up. In other words: I personally have no complain with audio sync between satellite receivers.

**Comment by Slashbunny** [comment]
FWIW, I have had no major issues using myMPD with a `mpd` proxy database setup for about 8 months. The only issue is the myMPD cache, but that is partly to do with the amount of music I have, not _just_ the database proxy. The initial album cache scan is very slow and the lack of persistent cache was painful (until the aforementioned patch), but I rarely power off the device (Intel Atom N230 CPU), so this is not showstopper. myMPD has the music directory set to "None". 

In order to get around playlists not working in satellite mode, I simply mounted the playlists directory over NFS from the same NAS to a local path and set that as the playlist directory for the local instance of `mpd`. This allows all instances to share the same playlists. 

The important parts of the configuration of a satellite `mpd` instance look like this, assuming the NAS holds your music as well as an instance of `mpd` maintaining the shared database:

```ini
...

music_directory "nfs://<your NAS host>/<path to music>"

database {
    plugin "proxy"
    host "<your NAS host>"
    password "<NAS mpd password>"
    keepalive "yes"
}

...
```

For reference, I have 900k+ files in the "primary" `mpd` database on the NAS. I have several `mpd` instances in satellite/proxy database mode with the music accessed over a NFS share (including Windows desktops/laptops). The standalone players are headless with a DAC and speakers and are configured using `mpd`'s [bit-perfect guidelines](https://mpd.readthedocs.io/en/stable/user.html#bit-perfect-playback). 

This would be more challenging to do with PulseAudio, since PulseAudio for Windows doesn't support RTP and getting bit-perfect playback with PulseAudio is much more fiddly (IMO).

**Reply by jcorporation** [reply]
> In order to get around playlists not working in satellite mode, I simply mounted the playlists directory over NFS from the same NAS to a local path and set that as the playlist directory for the local instance of `mpd`. This allows all instances to share the same playlists.

Great idea, I will add this to the new mpd satellite setup with myMPD documentation.

**Comment by jcorporation** [comment]
myMPD v13 adds a feature to maintain the stickers on a central MPD server.

**Comment by GioF71** [comment]
Hello, after a long time, I just wanted to share my current solution.
I wanted to avoid any file sharing like samba and nfs, so I ended up using autofs and sshfs.
Using docker for both mympd and mpd, it has been relatively easy to setup the satellites with matching paths for music directories.
And also, I noticed that all the operations are faster now, I don't exactly know since which release (probably v13 as you mentioned in the previous port). Maybe it has to do with the caching on each instance of mpd.
Eventually, in order to avoid sshfs as well, I might choose to host the files in a http server, but my concern is that this solution might slow down the master instance of mpd as well, because that instance should access the files through the http server in order to avoid file/directory name mismatch between server and satellites.
Do you @JuergenMang think that this solution would be much slower? Currently the master instance has direct file access to the library.

Thank you!

**Comment by jcorporation** [comment]
> Do you @JuergenMang think that this solution would be much slower? Currently the master instance has direct file access to the library.

You should test it.

**Comment by GioF71** [comment]
Hello, just tried that solution. It does not seem viable, the scanning was so slow that I had to gave up.
But I am not definitely sure, I need a clean try: today I just switched the local folders for the http server. I didn't clean the database first, I could have done it because I did a backup, but I just forgot. 
I will update this with further observations.
